### ISSUE TITLE
Run thumbnail generators in a different thread

### ISSUE THREAD

**Issue by martinsik** [kivy-garden/garden.filechooserthumbview#6]
I think many performance problems with generators could be solved easily by running them in threads at background not blocking the main app thread and then switching thumbnails with default icons.

**Comment by Davideddu** [comment]
Can you make the change and send me a pull request?

**Comment by Davideddu** [comment]
I checked the code, and the only way I think the viewers could be started in a different thread (without making things overly complicated) is using fifos, which I don't think we can use on Windows. Also, that way images are not going to be cached for further reuse.
In any case, this method is only going to add trouble. Kivy *may* hang after opening a FIFO if nobody's writing into it (because it's been already written earlier).

I switched to AsyncImage anyways, should speed up loading for large thumbnails.

**Comment by martinsik** [comment]
Hi, I'm not sure why you need fifo and why it shouldn't work under Windows. Maybe both of us mean something different.
I'll try the new version as you say and eventually try to make the threaded version by myself.

**Comment by Davideddu** [comment]
It would become too complicated to add a placeholder to the image source, add a callback to the thread to replace the placeholder with the actual image.
With fifos you would create a fifo and tell kivy to read it, then start the thumbnailer thread and feed it to the thumbnailer for writing.

**Comment by martinsik** [comment]
I sent a pull request kivy-garden/garden.filechooserthumbview#7 my changes are quite simple https://github.com/kivy-garden/garden.filechooserthumbview/pull/7/files.
In `FileChooserThumbView` I overloaded `_create_entry_widget()` to add context, widget and generator function to a queue that is run in another thread in `ThreadedThumbnailGenerator` class.

Check it out here: http://i.imgur.com/B3sA2D9.gif

Another very simple modification would be to add size argument to `get_png_from_video()` (and `-vf scale=xxx` to `ffmpeg`) because by default this generates thumbnails in the same resolution as the video. I tested it and it takes 500ms to make thumbnail for 720p video in full resolution while scaled down to 128 or 64px it took about 350ms.

If this works well I'll try to add more 2-4 threads because I think it'll speed up very well since `ffmpeg` spends a lot of time just loading the video file from disk and then rendering thumbnail which can run well in parallel. I'm not even sure if `ffmpeg` uses multiple cores when generating previews.
